### PR TITLE
Precompile regex patterns in BasicPasswordComplexityRule

### PR DESCRIPTION
**Pull request by mandar1045**
Follows up on #27038.

While looking at the new password complexity code I noticed the `validate()` method calls `String.matches()` four times — once for each character class check. The problem is `String.matches()` recompiles the regex on every invocation, so every login attempt or password change triggers four fresh `Pattern.compile()` calls under the hood.

Switched to static pre-compiled `Pattern` constants and `find()` instead of wrapping everything in `.*...*`. Nothing about the validation behavior changes — same passwords pass, same ones fail, same error messages. Just faster internally.

Also added a few tests that were missing:
- `getViolations()` on `PasswordComplexityException` was introduced in #27038 but had no direct test — added one
- edge cases around `minimumLength = 0`, negative values, and exact boundary length

### Testing done

Ran existing tests to make sure nothing broke. Added 4 new unit tests in `BasicPasswordComplexityRuleTest` covering the above.

### Screenshots (UI changes only)

N/A

### Proposed changelog entries

- Improve performance of `BasicPasswordComplexityRule` by pre-compiling regex patterns

### Proposed changelog category

/label internal

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] UI changes do not introduce regressions when enforcing the current default rules of [Content Security Policy Plugin](https://plugins.jenkins.io/csp/).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/core-pr-reviewers

### CODE REVIEW

File: core/src/test/java/jenkins/security/BasicPasswordComplexityRuleTest.java [hunk 1]
```diff
@@ -83,6 +84,51 @@ void multipleViolationsReportedTogether() {
...
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("at least 8 characters")),
+                "violations list should contain length error");
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("uppercase letter")),
+                "violations list should contain uppercase error");
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("digit")),
+                "violations list should contain digit error");
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("special character")),
+                "violations list should contain special character error");
+        assertTrue(e.getViolations().size() == 4,
+                "violations list should contain exactly one entry per violated rule");
```
Comment (MarkEWaite) [inline core/src/test/java/jenkins/security/BasicPasswordComplexityRuleTest.java]: A hamcrest assertion provides a better failure message.  Since the class already uses hamcrest in other cases, let's use it here as well.
Needs an additional `import static org.hamcrest.Matchers.is;`

```suggestion
        assertThat("violations list should contain exactly one entry per violated rule",
                   e.getViolations().size(), is(4));
```

**Review by MarkEWaite — CHANGES_REQUESTED** [review]

**Comment by mandar1045** [comment]
Thanks for the review, @MarkEWaite! Good catch on the missing import for the assertion—I've added it in. I also addressed the newline behavior nuance that Copilot pointed out by pre-compiling the original `.*...\*` patterns and sticking with `.matches()` so the exact original logic is preserved. 

Everything is squashed down to a single commit now. Let me know if there's anything else!

File: core/src/test/java/jenkins/security/BasicPasswordComplexityRuleTest.java [hunk 2]
```diff
@@ -83,6 +84,51 @@ void multipleViolationsReportedTogether() {
...
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("at least 8 characters")),
+                "violations list should contain length error");
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("uppercase letter")),
+                "violations list should contain uppercase error");
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("digit")),
+                "violations list should contain digit error");
+        assertTrue(e.getViolations().stream().anyMatch(v -> v.contains("special character")),
+                "violations list should contain special character error");
+        assertThat("violations list should contain exactly one entry per violated rule",
+                   e.getViolations().size(), is(4));
```
Comment (MarkEWaite) [inline core/src/test/java/jenkins/security/BasicPasswordComplexityRuleTest.java]: With further investigation, I found a Hamcrest assertion that matches the contents of a list without concern for the order of the list.  This assures that the list contains exactly those 4 violations without concern for the order of the violations in the list.

Needs `import static org.hamcrest.Matchers.containsInAnyOrder;`

Allows removal of the import of `assertTrue` and provides a clear failure message when the assertion is not satisfied.
```suggestion
        assertThat(e.getViolations(),
                   containsInAnyOrder("Password must be at least 8 characters long.",
                                      "Password must contain at least one uppercase letter (A-Z).",
                                      "Password must contain at least one digit (0-9).",
                                      "Password must contain at least one special character (e.g. !@#$%^&*)."));
```

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Use the Hamcrest assertion that checks contents of a list while ignoring ordering in the list.

**Comment by MarkEWaite** [comment]
> Everything is squashed down to a single commit now. Let me know if there's anything else!

Thanks.  Please don't rebase commits in this repository.  It risks separating the commit comments from the source code.  We usually squash merge pull requests in this repository, so we don't need to rebase prior to the final merge

**Comment by mandar1045** [comment]
Ah, apologies for the rebase! I'll stick to standard commits from here on out so the review history isn't disrupted. I've updated the test to use `containsInAnyOrder` as suggested—it's much cleaner, thanks!

File: core/src/test/java/jenkins/security/BasicPasswordComplexityRuleTest.java [hunk 3]
```diff
@@ -83,6 +85,46 @@ void multipleViolationsReportedTogether() {
...
+    void violationsListContainsIndividualMessages() {
+        BasicPasswordComplexityRule rule = new BasicPasswordComplexityRule(8, true, true, true, true);
+
+        PasswordComplexityException e = assertThrows(PasswordComplexityException.class, () -> rule.validate("bad"));
+        // Violations list should have exactly these 4 violations in any order
+        assertThat(e.getViolations(),
+                   containsInAnyOrder("Password must be at least 8 characters long.",
+                                      "Password must contain at least one uppercase letter (A-Z).",
+                                      "Password must contain at least one digit (0-9).",
+                                      "Password must contain at least one special character (e.g. !@#$%^&*)."));
```
[the remark this answers is not in the document]
Comment (MarkEWaite) [reply core/src/test/java/jenkins/security/BasicPasswordComplexityRuleTest.java:97]: I find the literal strings easier to read than the message references.  I think that the test is clearer with literal strings than with references to the messages properties.  I acknowledge that is a personal preference.  I'm not ready to mandate that literal strings must be used, just stating my personal preference.
